File: pkg-r/tests/testthat/test-remote-opentargets.R
# The Open Targets resolver: the first GraphQL (POST) remote check. The test
# transport for POST receives the request body and reads the id from it.

.ot_post_stub <- function(covered) {
  function(url, body, timeout) {
    id <- jsonlite::fromJSON(body)$variables$ensemblId
    target <- if (id %in% covered) list(id = id) else NULL
    list(status = 200, body = list(data = list(target = target)))
  }
}

test_that("a covered Open Targets target is valid", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport_post = .ot_post_stub("ENSG00000139618")
  )
  res <- check_id("ENSG00000139618", source_db = "opentargets", how = "remote")
  expect_true(res$valid)
  expect_identical(res$normalized, "ENSG00000139618")
})

test_that("a gene the platform does not cover is absent", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport_post = .ot_post_stub(character(0))
  )
  res <- check_id("ENSG00000000000", source_db = "opentargets", how = "remote")
  expect_false(res$valid)
})

test_that("a malformed case suggests the covered form", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport_post = .ot_post_stub("ENSG00000139618")
  )
  res <- check_id("ensg00000139618", source_db = "opentargets", how = "remote")
  expect_false(res$valid)
  expect_identical(res$suggestion, "ENSG00000139618")
})

test_that("an unexpected status is left indeterminate under on_error", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport_post = function(url, body, timeout) {
      list(status = 500, body = NULL)
    }
  )
  res <- check_id(
    "ENSG00000139618",
    source_db = "opentargets",
    how = "remote",
    on_error = "indeterminate"
  )
  expect_true(is.na(res$valid))
  expect_false(is.na(res$error))
})

test_that("opentargets is a registered source", {
  info <- source_info()
  row <- info[info$key == "opentargets", ]
  expect_identical(nrow(row), 1L)
  expect_identical(row$modes, "pattern, remote")
})

File: pkg-py/tests/test_remote_opentargets.py
"""The Open Targets resolver: the first GraphQL (POST) remote check."""

import json

import biobouncer
import biobouncer._remote as remote


def _stub(covered):
    """A GraphQL _http_post stub: ids in ``covered`` resolve to a target."""

    def _http_post(url, data, timeout=30):
        assert url.endswith("/graphql")
        ident = json.loads(data)["variables"]["ensemblId"]
        target = {"id": ident} if ident in covered else None
        return 200, {"data": {"target": target}}

    return _http_post


def test_covered_target_is_valid(monkeypatch, tmp_path):
    monkeypatch.setenv("BIOBOUNCER_CACHE_DIR", str(tmp_path))
    monkeypatch.setattr(remote, "_http_post", _stub({"ENSG00000139618"}))
    res = biobouncer.check_id("ENSG00000139618", source_db="opentargets", how="remote")[
        0
    ]
    assert res.valid is True
    assert res.normalized == "ENSG00000139618"


def test_uncovered_gene_is_absent(monkeypatch, tmp_path):
    monkeypatch.setenv("BIOBOUNCER_CACHE_DIR", str(tmp_path))
    monkeypatch.setattr(remote, "_http_post", _stub(set()))
    res = biobouncer.check_id("ENSG00000000000", source_db="opentargets", how="remote")[
        0
    ]
    assert res.valid is False


def test_malformed_case_suggests_the_covered_form(monkeypatch, tmp_path):
    monkeypatch.setenv("BIOBOUNCER_CACHE_DIR", str(tmp_path))
    monkeypatch.setattr(remote, "_http_post", _stub({"ENSG00000139618"}))
    res = biobouncer.check_id("ensg00000139618", source_db="opentargets", how="remote")[
        0
    ]
    assert res.valid is False
    assert res.suggestion == "ENSG00000139618"


def test_body_carries_the_graphql_query_and_id():
    body = remote._opentargets_body(None, "ENSG00000139618")
    payload = json.loads(body)
    assert "target(ensemblId:" in payload["query"]
    assert payload["variables"] == {"ensemblId": "ENSG00000139618"}


def test_unexpected_status_is_indeterminate(monkeypatch, tmp_path):
    monkeypatch.setenv("BIOBOUNCER_CACHE_DIR", str(tmp_path))

    def _boom(url, data, timeout=30):
        return 500, None

    monkeypatch.setattr(remote, "_http_post", _boom)
    res = biobouncer.check_id(
        "ENSG00000139618",
        source_db="opentargets",
        how="remote",
        on_error="indeterminate",
    )[0]
    assert res.valid is None
    assert res.error is not None


def test_opentargets_is_a_registered_source():
    info = {row["key"]: row for row in biobouncer.source_info()}
    assert "opentargets" in info
    assert info["opentargets"]["modes"] == ["pattern", "remote"]
